Examples/GraphColoring.lean

```diff
@@ -0,0 +1,232 @@
+import LeanSAT.Model.PropFun
+import LeanSAT.Encode.EncCNF
+import LeanSAT.Encode.VEncCNF
+import LeanSAT.Solver.Basic
+import LeanSAT.Solver.Dimacs
+import LeanSAT.Data.ICnf
+
+import Mathlib.Data.Fintype.Prod
+import Mathlib.Data.Prod.Lex
+import Mathlib.Data.Sum.Order
+import Mathlib.Combinatorics.SimpleGraph.Basic
+import Mathlib.Tactic
+
+namespace Array
+
+def finRange (n : Nat) : Array (Fin n) :=
+  ⟨List.finRange n⟩
+
+@[simp] theorem mem_finRange {n} (x : Fin n) : x ∈ finRange n := by simp [finRange, mem_def]
+@[simp] theorem finRange_data (n) : (Array.finRange n).data = List.finRange n := rfl
+
+end Array
+
+-------------------------------------------------------------------
+
+open LeanSAT Model PropFun
+
+-- A graph is a symmetric function from two vertices to a boolean
+-- The number of vertices `n` is specified ahead of time
+abbrev Graph (n : Nat) := SimpleGraph (Fin n)
+
+def Coloring (n : Nat) (numColors : Nat) := (Fin n) → (Fin numColors)
+
+def isValidColoring (G : Graph n) (C : Coloring n c) :=
+  ∀ ⦃u v⦄, G.Adj u v → C u ≠ C v
+
+-- Vertex variables involved in graph coloring
+inductive ColorVars (n : Nat)
+  | blue   (v : Fin n)
+  | red    (v : Fin n)
+  | green  (v : Fin n)
+  | yellow (v : Fin n)
+  deriving DecidableEq, Repr
+
+instance : FinEnum (ColorVars n) := .ofEquiv {
+  toFun := fun c => match c with
+    | .blue   v => ((⟨0, by decide⟩ : Fin 4), v)
+    | .red    v => ((⟨1, by decide⟩ : Fin 4), v)
+    | .green  v => ((⟨2, by decide⟩ : Fin 4), v)
+    | .yellow v => ((⟨3, by decide⟩ : Fin 4), v)
+  invFun := fun (c, v) =>
+    match hc : c with
+    | ⟨0, _⟩ => .blue v
+    | ⟨1, _⟩ => .red v
+    | ⟨2, _⟩ => .green v
+    | ⟨3, _⟩ => .yellow v
+    | ⟨n + 4, hn⟩ => by contradiction
+  left_inv := sorry
+  right_inv := sorry
+}
+
+-- Open the namespace so we can use ".blue" instead of "Coloring.blue", etc.
+open ColorVars
+
+-- We sometimes need to provide a lexicographic ordering for our variables
+instance : LinearOrder (ColorVars n) :=
+  LinearOrder.lift'
+    (β := Fin n ⊕ₗ Fin n ⊕ₗ Fin n ⊕ₗ Fin n)
+    (fun
+      | .blue v => .inlₗ v
+      | .red v => .inrₗ <| .inlₗ v
+      | .green v => .inrₗ <| .inrₗ <| .inlₗ v
+      | .yellow v => .inrₗ <| .inrₗ <| .inrₗ v)
+    (by rintro ⟨⟩ ⟨⟩ <;> simp)
+
+-- Given a vertex, we should expect that it gets a color
+def eachVertexGetsAColor (u : Fin n) : PropFun (ColorVars n) :=
+  [propfun|
+    ({blue   u}  ∨
+     {red    u}  ∨
+     {green  u}  ∨
+     {yellow u})
+  ]
+
+-- Alternative definition treating CNF clauses as boolean algebras
+def eachVertexGetsAColor' (v : Fin n) : PropFun (ColorVars n) :=
+  (blue v) ⊔ (red v) ⊔ (green v) ⊔ (yellow v)
```

Comment:
I think I prefer this notation. Maybe we should standardize around this notation? (and remove the other from this example)

We could even add it as syntax for PropForm, even though PropForm is not a booleanalgebra

Reply:
created #17